### ISSUE TITLE
[BUG] Test Explorer Groups test incorrect

### ISSUE THREAD

**Issue by MikeSchulze** [microsoft/vscode-dotnettools#841]
### Describe the Issue

I have written my own test adapter to discover GdUnit4 API tests.
All tests are detected and can be executed and debugged, which is good.

My test API supports parameterized tests like
```
        [TestCase(1, 2, 3, 6, TestName = "TestCaseA")]
        [TestCase(3, 4, 5, 12, TestName = "TestCaseB")]
        [TestCase(6, 7, 8, 21, TestName = "TestCaseC")]
        public void TestCasesWithCustomTestName(int a, double b, int c, int expect)
        {
            AssertThat(a + b + c).IsEqual(expect);
        }
```
All tests are discovered, but the explorer does a wrong grouping.
For example, I have a test suite:
```

namespace GdUnit4.Tests.Core
{
    using static Assertions;
    using static Utils;

    [TestSuite]
    public class ExampleTestSuite
    {
        [Before]
        public void Before()
        {
            // GD.PrintS("calling Before");
        }

        [After]
        public void After()
        {
            //GD.PrintS("calling After");
        }

        [BeforeTest]
        public void BeforeTest()
        {
            //GD.PrintS("calling BeforeTest");
        }

        [AfterTest]
        public void AfterTest()
        {
            // GD.PrintS("calling AfterTest");
        }

        [TestCase]
        public void TestFoo()
        {
            AssertBool(true).IsEqual(true);
        }

        [TestCase]
        public void TestBar()
        {
            AssertBool(true).IsEqual(true);
        }

        [TestCase]
        public async Task Waiting()
        {
            await DoWait(200);
        }

        [TestCase]
        public void TestFooBar()
        {
            AssertBool(true).IsEqual(true);
        }

        [TestCase(1, 2, 3, 6)]
        [TestCase(3, 4, 5, 12)]
        [TestCase(6, 7, 8, 21)]
        public void TestCaseArguments(int a, int b, int c, int expect)
        {
            AssertThat(a + b + c).IsEqual(expect);
        }

        [TestCase(1, 2, 3, 6, TestName = "TestCaseA")]
        [TestCase(3, 4, 5, 12, TestName = "TestCaseB")]
        [TestCase(6, 7, 8, 21, TestName = "TestCaseC")]
        public void TestCasesWithCustomTestName(int a, double b, int c, int expect)
        {
            AssertThat(a + b + c).IsEqual(expect);
        }
    }
}
```
The test explorer shows them sliced into

![image](https://github.com/microsoft/vscode-dotnettools/assets/347037/7cd8dc1d-55a8-411a-87c3-4a455e71fd4b)

I think I generate the test-case `FullyQualifiedName` and `TestName` attributes correctly.
I tested against `.NET TEST EXPLORER` and it shows the test suite in one group.

![image](https://github.com/microsoft/vscode-dotnettools/assets/347037/ebb1273c-1245-4062-a69b-4c28ee74e581)

### Steps To Reproduce

My test adapter is not yet published, so I can't offer any steps to reproduce it.

My test discover is building test cases by.
```
TestCase testCase = new(fullyQualifiedName, new Uri(GdUnit4TestExecutor.ExecutorUri), assemblyPath)
        {
            DisplayName = testName,
            FullyQualifiedName = fullyQualifiedName,
            CodeFilePath = navData.Source,
            LineNumber = navData.Line
        };
```
The `DisplayName` is e.g. `TestCaseArguments(1, 2, 3, 6)` and the `FullyQualifiedName` contains the full namespace
e.g. `GdUnit4.Tests.Core.ExampleTestSuite.TestCaseArguments(1, 2, 3, 6)`

Maybe I'm doing something wrong here, the official test adapter documentation is very poor on this. There is no explanation of how to set the attributes and how to group them. Maybe there are special custom attributes that need to be set that I am not aware of.
I would be grateful for any help.

### Expected Behavior

It should be grouped like in the `.NET TEST EXPLORER` 

![image](https://github.com/microsoft/vscode-dotnettools/assets/347037/ebb1273c-1245-4062-a69b-4c28ee74e581)

### Environment Information

Windows 11
-----------------------
Visual Studio Code
```
Version: 1.85.1 (user setup)
Commit: 0ee08df0cf4527e40edc9aa28f4b5bd38bbff2b2
Date: 2023-12-13T09:49:37.021Z
Electron: 25.9.7
ElectronBuildId: 25551756
Chromium: 114.0.5735.289
Node.js: 18.15.0
V8: 11.4.183.29-electron.0
OS: Windows_NT x64 10.0.22631
```
--------------
global,json
```
{
    "sdk": {
        "version": "7.0.0-rc",
        "rollForward": "latestMinor"
    },
    "msbuild-sdks": {
        "Godot.NET.Sdk": "4.2.0"
    }
}
```
Project includes
```

    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.8.0" />
    <PackageReference Include="Microsoft.TestPlatform.ObjectModel" Version="17.8.0" PrivateAssets="All" />
    <PackageReference Include="coverlet.collector" Version="6.0.0" />
    <PackageReference Include="Newtonsoft.Json" Version="13.0.3" />
 ```
 .runsettings
 
 ```
 <?xml version="1.0" encoding="utf-8"?>
<RunSettings>
    <RunConfiguration>
        <MaxCpuCount>1</MaxCpuCount>
        <ResultsDirectory>./TestResults</ResultsDirectory>
        <TargetFrameworkVersion>net7.0</TargetFrameworkVersion>
        <TestSessionTimeout>180000</TestSessionTimeout>
        <TreatNoTestsAsError>true</TreatNoTestsAsError>
    </RunConfiguration>

    <LoggerRunSettings>
        <Loggers>
            <Logger friendlyName="console" enabled="True">
                <Configuration>
                    <Verbosity>detailed</Verbosity>
                </Configuration>
            </Logger>
        </Loggers>
    </LoggerRunSettings>

    <TestRunParameters>
        <Parameter name="TEST_PARAM" value="test" />
    </TestRunParameters>

    <GdUnit4>
        <!-- Additonal Godot runtime parameters-->
        <Parameters></Parameters>
    </GdUnit4>
</RunSettings>
```

**Comment by peterwald** [comment]
There are various ways to control how a test is displayed. The `FullyQualifiedName+DisplayName` is the primary way, but there are all sorts of heuristics that evolved over time due to compatibility and evolution of the APIs. I believe you might get what you want if you remove the parenthetical portion from the end of the FullyQualifiedName (a.k.a. FQN) and make it just the MethodName (or DisplayName if you want to make it different from the method name).  The FQN should not have to be unique to more than the method.

If you want a more modern way of more precisely determining what is displayed, you can add a Hierarchy property. [See this document for details](https://github.com/microsoft/vstest/blob/main/docs/RFCs/0033-Hierarchy-TestCase-Property.md)

There is a lot of other documentation in that same repo that might help answer other questions you may have.

**Comment by MikeSchulze** [comment]
@peterwald I switched to using the managed properties, but it ends up in the same result.
```
 testCase.SetPropertyValue(ManagedTypeProperty, descriptor.ManagedType);
 testCase.SetPropertyValue(ManagedMethodProperty, descriptor.ManagedMethod);
```

Maybe I have a clash in the name building?
```
Discovered TestCase 'TestFoo':
    FullyQualifiedName: 'GdUnit4.Tests.Core.ExampleTestSuite.TestFoo'
    ManagedType: 'GdUnit4.Tests.Core.ExampleTestSuite'
    ManagedMethod: 'TestFoo'
    HierarchyValues ''

...
Discovered TestCase 'TestCaseArguments:
    FullyQualifiedName: 'GdUnit4.Tests.Core.ExampleTestSuite.TestCaseArguments.TestCaseArguments(1, 2, 3, 6)'
    ManagedType: 'GdUnit4.Tests.Core.ExampleTestSuite'
    ManagedMethod: 'TestCaseArguments.TestCaseArguments(1, 2, 3, 6)'
    HierarchyValues '[<Null>, "GdUnit4.Tests.Core", "ExampleTestSuite", "TestCaseArguments(System.Int32,System.Int32,System.Int32,System.Int32)"]'

....

Discovered TestCase 'TestCasesWithCustomTestName:
    FullyQualifiedName: 'GdUnit4.Tests.Core.ExampleTestSuite.TestCasesWithCustomTestName.TestCaseA(1, 2, 3, 6)'
    ManagedType: 'GdUnit4.Tests.Core.ExampleTestSuite'
    ManagedMethod: 'TestCasesWithCustomTestName.TestCaseA(1, 2, 3, 6)'
    HierarchyValues '[<Null>, "GdUnit4.Tests.Core", "ExampleTestSuite", "TestCasesWithCustomTestName(System.Int32,System.Double,System.Int32,System.Int32)"]'
```

The tests are defined as
```
    [TestCase]
    public void TestFoo()
        => AssertBool(true).IsEqual(true);

   
    [TestCase(1, 2, 3, 6)]
    [TestCase(3, 4, 5, 12)]
    [TestCase(6, 7, 8, 21)]
    public void TestCaseArguments(int a, int b, int c, int expect)
        => AssertThat(a + b + c).IsEqual(expect);

    [TestCase(1, 2, 3, 6, TestName = "TestCaseA")]
    [TestCase(3, 4, 5, 12, TestName = "TestCaseB")]
    [TestCase(6, 7, 8, 21, TestName = "TestCaseC")]
    public void TestCasesWithCustomTestName(int a, double b, int c, int expect)
    ````

Any hints what I do wrong?
Is it correct to create a separate test case for each parameterized test?
e.g. for `TestCasesWithCustomTestName` there are three test cases of the type `Microsoft.VisualStudio.TestPlatform.ObjectModel.TestCase`.

**Comment by peterwald** [comment]
> Is it correct to create a separate test case for each parameterized test?

Yes. You should create a new test for each parameterized instance of a test.

ManagedMethod must follow the spec in order to be parsed correctly. A method name should not include a period (unless it were escaped). I suggest you use the [Microsoft.TestPlatform.AdapterUtilities](https://www.nuget.org/packages/Microsoft.TestPlatform.AdapterUtilities) nuget package as it has [helpers](https://github.com/microsoft/vstest/tree/main/src/Microsoft.TestPlatform.AdapterUtilities/ManagedNameUtilities) to create proper ManagedType and ManagedMethod identifiers. It also has a helper to generate a default for `HierarchyValues` and a parser that can be used to verify the identifier.

Your properties should look something more like this.

```
Discovered TestCase 'TestFoo':
    FullyQualifiedName: 'GdUnit4.Tests.Core.ExampleTestSuite.TestFoo'
    ManagedType: 'GdUnit4.Tests.Core.ExampleTestSuite'
    ManagedMethod: 'TestFoo()'
    HierarchyValues: [<Null>, "GdUnit4.Tests.Core", "ExampleTestSuite", "TestFoo"]
    DisplayName: "TestFoo"

...
Discovered TestCase 'TestCaseArguments:
    FullyQualifiedName: 'GdUnit4.Tests.Core.ExampleTestSuite.TestCaseArguments.TestCaseArguments(1, 2, 3, 6)'
    ManagedType: 'GdUnit4.Tests.Core.ExampleTestSuite'
    ManagedMethod: 'TestCaseArguments(System.Int32,System.Int32,System.Int32,System.Int32)'
    HierarchyValues: [<Null>, "GdUnit4.Tests.Core", "ExampleTestSuite", "TestCaseArguments"]
    DisplayName: 'TestCaseArguments(1,2,3,6)'

....

Discovered TestCase 'TestCasesWithCustomTestName:
    FullyQualifiedName: 'GdUnit4.Tests.Core.ExampleTestSuite.TestCasesWithCustomTestName.TestCaseA(1, 2, 3, 6)'
    ManagedType: 'GdUnit4.Tests.Core.ExampleTestSuite'
    ManagedMethod: 'TestCasesWithCustomTestName(System.Int32,System.Double,System.Int32,System.Int32)'
    HierarchyValues: [<Null>, "GdUnit4.Tests.Core", "ExampleTestSuite", "TestCasesWithCustomTestName"]
    DisplayName: "TestCaseA"
```

**Comment by MikeSchulze** [comment]
Thanks for your answer, I actually use the [AdapterUtilities](https://www.nuget.org/packages/Microsoft.TestPlatform.AdapterUtilities) in this way
```
   ManagedNameHelper.GetManagedName(mi, out var managedType, out var managedMethod, out var hierarchyValues);
   ManagedNameParser.ParseManagedMethodName(managedMethod, out var methodName, out var arity, out var parameterTypes);
   hierarchyValues[HierarchyConstants.Levels.ContainerIndex] = null;
   hierarchyValues[HierarchyConstants.Levels.TestGroupIndex] = methodName;
```
and create the testcase by:
```
    private TestCase BuildTestCase(TestCaseDescriptor descriptor, string assemblyPath, CodeNavigation navData, GdUnit4Settings gdUnitSettings)
    {
        TestCase testCase = new(descriptor.FullyQualifiedName, new Uri(GdUnit4TestExecutor.ExecutorUri), assemblyPath)
        {
            DisplayName = descriptor.DisplayName,//BuildDisplayName(descriptor.DisplayName, descriptor.FullyQualifiedName, gdUnitSettings),
            FullyQualifiedName = descriptor.FullyQualifiedName,
            CodeFilePath = navData.Source,
            LineNumber = navData.Line
        };

        testCase.SetPropertyValue(TestCaseNameProperty, descriptor.DisplayName);
        testCase.SetPropertyValue(ManagedTypeProperty, descriptor.ManagedType);
        testCase.SetPropertyValue(ManagedMethodProperty, descriptor.ManagedMethod);
        if (descriptor.HierarchyValues?.Count > 0)
        {
            testCase.SetPropertyValue(HierarchyProperty, descriptor.HierarchyValues?.ToArray<string?>());
        }
        return testCase;
    }
```
It results into:
```
Discovered TestCase:
    Id: '79b54e28-f6c4-d9a1-1bd4-cc70ab6d2bd6'
    DisplayName: 'TestCaseArguments(1, 2, 3, 6)'
    FullyQualifiedName: 'GdUnit4.Tests.Core.ExampleTestSuite.TestCaseArguments(1, 2, 3, 6)'
    ExecutorUri: 'executor://gdunit4.testadapter/'
    Source: 'D:\development\workspace\gdUnit4Mono\test\.godot\mono\temp\bin\Debug\gdUnit4Test.dll'
    Properties: '["File Path", "Name", "Executor Uri", "FullyQualifiedName", "Id", "Line Number", "Source", "Hierarchy", "ManagedType", "Name", "ManagedMethod"]'
    ManagedType: 'GdUnit4.Tests.Core.ExampleTestSuite'
    ManagedMethod: 'TestCaseArguments(System.Int32,System.Int32,System.Int32,System.Int32)'
    HierarchyValues: '[<Null>, "GdUnit4.Tests.Core", "ExampleTestSuite", "TestCaseArguments"]'

Discovered TestCase:
    Id: '79b54e28-f6c4-d9a1-1bd4-cc70ab6d2bd6'
    DisplayName: 'TestCaseArguments(3, 4, 5, 12)'
    FullyQualifiedName: 'GdUnit4.Tests.Core.ExampleTestSuite.TestCaseArguments(3, 4, 5, 12)'
    ExecutorUri: 'executor://gdunit4.testadapter/'
    Source: 'D:\development\workspace\gdUnit4Mono\test\.godot\mono\temp\bin\Debug\gdUnit4Test.dll'
    Properties: '["File Path", "Name", "Executor Uri", "FullyQualifiedName", "Id", "Line Number", "Source", "Hierarchy", "ManagedType", "Name", "ManagedMethod"]'
    ManagedType: 'GdUnit4.Tests.Core.ExampleTestSuite'
    ManagedMethod: 'TestCaseArguments(System.Int32,System.Int32,System.Int32,System.Int32)'
    HierarchyValues: '[<Null>, "GdUnit4.Tests.Core", "ExampleTestSuite", "TestCaseArguments"]'

Discovered TestCase:
    Id: '79b54e28-f6c4-d9a1-1bd4-cc70ab6d2bd6'
    DisplayName: 'TestCaseArguments(6, 7, 8, 21)'
    FullyQualifiedName: 'GdUnit4.Tests.Core.ExampleTestSuite.TestCaseArguments(6, 7, 8, 21)'
    ExecutorUri: 'executor://gdunit4.testadapter/'
    Source: 'D:\development\workspace\gdUnit4Mono\test\.godot\mono\temp\bin\Debug\gdUnit4Test.dll'
    Properties: '["File Path", "Name", "Executor Uri", "FullyQualifiedName", "Id", "Line Number", "Source", "Hierarchy", "ManagedType", "Name", "ManagedMethod"]'
    ManagedType: 'GdUnit4.Tests.Core.ExampleTestSuite'
    ManagedMethod: 'TestCaseArguments(System.Int32,System.Int32,System.Int32,System.Int32)'
    HierarchyValues: '[<Null>, "GdUnit4.Tests.Core", "ExampleTestSuite", "TestCaseArguments"]'
```
And is shown in the explorer 
![image](https://github.com/microsoft/vscode-dotnettools/assets/347037/753f26f5-6257-4873-859a-230859f7933d)


What I see the generated testcase ID is the same 
`Id: '79b54e28-f6c4-d9a1-1bd4-cc70ab6d2bd6'`

When I replace the default generated ID by using the `TestIdProvider` each test case has a unique ID and it results into 
![image](https://github.com/microsoft/vscode-dotnettools/assets/347037/16c0c3fd-a39c-4412-8aaa-3efa63236339)

So it looks like the default ID generation has a problem.

Thanks for your help, that forces me in the right direction now ;)
The question now is whether the default generated test ID has an error, or whether it has to be created manually, I can't find anything about this in the documentation.

**Comment by peterwald** [comment]
The default test id generation uses the Executor/Source/FullyQualifiedName to hash to a TestCaseId for legacy tests. However, when providing a `ManagedType`/`ManagedMethod`, [it uses those to get the string](https://github.com/microsoft/vstest/blob/e446f0eea6e51f883172abaca649b4d6484a08e5/src/Microsoft.TestPlatform.ObjectModel/TestCase.cs#L310) for creating the id. Since those are the same for each data-driven test case, that is why you get the same id.

When providing the `ManagedType`/`ManagedMethod`, we expect a unique and deterministic (the id is always the same as long as the test is the "same") id to be provided. That is the meaning behind [this paragraph](https://github.com/microsoft/vstest/blob/main/docs/RFCs/0017-Managed-TestCase-Properties.md#uniqueness) in the spec. I can see that it might not be clear enough.

In short, when providing the `ManagedType`/`ManagedMethod` for a test case you should also use the `TestIdProvider` and append representations of your data parameters (as string or bytes), so that each test case has a unique and deterministic id. If for some reason you have duplicate data driven test cases, you would need to add some additional factor to make them unique.

**Comment by MikeSchulze** [comment]
Thank you very much for your reply, I was able to solve my problem with all the new information.
Thank you again, you were really very helpful.
💯
